### PR TITLE
feat(v2): implement client modules API

### PR DESCRIPTION
**Pull request by yangshun**
## Motivation

Implement client modules API for plugins to define what modules they want to import on the client.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Unit tests. I tested locally with some client modules but didn't commit it. Next step is to import Infima as a client module.

## Related PRs

NA.

### CODE REVIEW

File: packages/docusaurus/src/client/App.js [hunk 1]
```diff
@@ -8,12 +8,15 @@
...
 
-import Head from '@docusaurus/Head';
 import routes from '@generated/routes';
 import siteConfig from '@generated/docusaurus.config';
+
+import Head from '@docusaurus/Head';
 import DocusaurusContext from '@docusaurus/context';
 import PendingNavigation from './PendingNavigation';
 
+require('@generated/client-modules');
```
Comment (endiliey) [inline packages/docusaurus/src/client/App.js]: why is it require

Comment (yangshun) [reply packages/docusaurus/src/client/App.js]: Oh yeah I can make it `import`. But it's really the same now. Webpack turns all of them into its own require.

Comment (endiliey) [reply packages/docusaurus/src/client/App.js]: lets change it for consistency

Comment (endiliey) [reply packages/docusaurus/src/client/App.js]: easier to migrate to TS later on

**Pull request merged**